plugins/delete-by-query/src/main/java/org/elasticsearch/action/deletebyquery/TransportDeleteByQueryAction.java

```diff
@@ -0,0 +1,307 @@
+    }
+
+    class AsyncDeleteByQueryAction {
+
+        private final DeleteByQueryRequest request;
+        private final ActionListener<DeleteByQueryResponse> listener;
+
+        private final long startTime;
+        private final AtomicBoolean timedOut;
+        private final AtomicLong total;
+
+        private volatile ShardOperationFailedException[] shardFailures;
+        private final Map<String, IndexDeleteByQueryResponse> results;
+
+        AsyncDeleteByQueryAction(DeleteByQueryRequest request, ActionListener<DeleteByQueryResponse> listener) {
+            this.request = request;
+            this.listener = listener;
+            this.startTime = System.currentTimeMillis();
+            this.timedOut = new AtomicBoolean(false);
+            this.total = new AtomicLong(0L);
+            this.shardFailures = ShardSearchFailure.EMPTY_ARRAY;
+            this.results = new HashMap<>();
+        }
+
+        public void start() {
+            executeScan();
+        }
+
+        void executeScan() {
+            try {
+                final SearchRequest scanRequest = new SearchRequest(request.indices()).types(request.types()).indicesOptions(request.indicesOptions());
+                scanRequest.searchType(SearchType.SCAN).scroll(request.scroll());
+                if (request.routing() != null) {
+                    scanRequest.routing(request.routing());
+                }
+
+                SearchSourceBuilder source = new SearchSourceBuilder().query(request.source()).fields("_routing", "_parent").fetchSource(false).version(true);
+                if (request.size() != null) {
+                    source.size(request.size());
+                }
+                if (request.timeout() != null) {
+                    source.timeout(request.timeout());
+                }
+                scanRequest.source(source);
+
+                logger.trace("executing scan request");
+                searchAction.execute(scanRequest, new ActionListener<SearchResponse>() {
+                    @Override
+                    public void onResponse(SearchResponse searchResponse) {
+                        long hits = searchResponse.getHits().getTotalHits();
+                        logger.trace("scan request executed: found [{}] document(s) to delete", hits);
+                        addShardFailures(searchResponse.getShardFailures());
+
+                        if (hits == 0) {
+                            listener.onResponse(buildResponse());
+                            return;
+                        }
+                        total.set(hits);
+
+                        logger.trace("start scrolling [{}] document(s)", hits);
+                        executeScroll(searchResponse.getScrollId());
+                    }
+
+                    @Override
+                    public void onFailure(Throwable e) {
+                        listener.onFailure(e);
+                    }
+                });
+            } catch (Throwable t) {
```

Comment:
I think you should have such a catch clause in every method in here since might easily hang the entire requires if there is a simple assertion tripped elsewhere